File: gamebasic/modules/wifi.py
"""WiFi-Management-Modul fuer GameBasic (Windows-only).

Greift auf die Windows-eigene `netsh wlan`-CLI zu - keine externe
Python-Lib noetig. Linux/macOS werden *nicht* unterstuetzt (dort wuerde
man `nmcli` bzw. `networksetup` ansprechen, was eigene Parser braucht).

Built-ins:
    WIFI_AVAILABLE()              -> BOOLEAN  (true wenn Windows + WLAN-Adapter)
    WIFI_CURRENT()                -> STRING   (aktuelle SSID, "" wenn nicht verbunden)
    WIFI_SIGNAL()                 -> INTEGER  (0..100, -1 wenn nicht verbunden)
    WIFI_SCAN()                   -> STRING   (multi-line "ssid|signal_pct")
    WIFI_CONNECT(ssid$, pass$)    -> BOOLEAN  (Profil anlegen + connect-Befehl)
    WIFI_DISCONNECT()             -> BOOLEAN
    WIFI_PROFILES()               -> STRING   (multi-line gespeicherter Profile)
    WIFI_DELETE_PROFILE(name$)    -> BOOLEAN

WICHTIG fuer WIFI_CONNECT:
- Funktioniert fuer offene Netze (pass$ leer) und WPA2-PSK (am haeufigsten).
- WPA-Enterprise (RADIUS), WPA3, captive portals werden NICHT unterstuetzt.
- Gibt nur zurueck ob das connect-Kommando angenommen wurde - polle danach
  WIFI_CURRENT() um zu sehen ob die Verbindung wirklich steht.
- Passwoerter werden in Klartext im Windows-Profil-Store abgelegt (Standard
  von netsh). Fuer Skript-Verwendung okay, fuer Production ungeeignet.
"""
from __future__ import annotations

import re
import subprocess
import sys
import tempfile
from html import escape as _xml_escape
from pathlib import Path

from ..builtins_registry import builtin
from ..errors import GBRuntimeError

_IS_WINDOWS = sys.platform == "win32"

# Versteckt das schwarze Konsolenfenster waehrend netsh-Aufrufen
_CREATE_NO_WINDOW = 0x08000000 if _IS_WINDOWS else 0


def _ensure_platform():
    if not _IS_WINDOWS:
        raise GBRuntimeError(
            "Modul WIFI: nur unter Windows unterstuetzt (nutzt netsh wlan).\n"
            "Linux: 'nmcli' direkt aufrufen; macOS: 'networksetup'."
        )


def _decode(raw: bytes) -> str:
    """netsh schreibt je nach Windows-Version in UTF-8 oder im OEM-Codepage.

    Reihenfolge: utf-8 zuerst, weil moderne Windows-11-Systeme netsh-Output
    bereits als UTF-8 liefern. Single-Byte-Codepages wie cp850/cp1252 wuerden
    diese Bytes klaglos akzeptieren und Umlaute kaputt-dekodieren (z.B. "ä"
    aus den UTF-8-Bytes 0xC3 0xA4 wuerde unter cp850 zu "├ñ"). UTF-8
    schlaegt bei wirklichem OEM-Output sauber fehl, dann faellt der Decoder
    auf den OEM-Codepage zurueck.
    """
    for enc in ("utf-8", "oem", "cp1252", "latin-1"):
        try:
            return raw.decode(enc)
        except (UnicodeDecodeError, LookupError):
            continue
    return raw.decode("latin-1", errors="replace")


def _run_netsh(args: list) -> tuple:
    """(returncode, output_text). Faengt FileNotFound/Timeout sauber ab."""
    try:
        proc = subprocess.run(
            ["netsh"] + args,
            capture_output=True,
            timeout=15,
            creationflags=_CREATE_NO_WINDOW,
        )
    except FileNotFoundError:
        raise GBRuntimeError("WIFI: 'netsh' nicht gefunden")
    except subprocess.TimeoutExpired:
        raise GBRuntimeError("WIFI: 'netsh' Timeout")
    return proc.returncode, _decode(proc.stdout) + _decode(proc.stderr)


# Regex fuer Output-Parsing (lokalisierungs-tolerant: stuetzt sich auf
# stabile Stichworte SSID / Signal / State|Status).

# Match "SSID : foo" aber NICHT "BSSID : ..." dank \b
_RE_SSID_LINE = re.compile(r"(?m)^\s*\bSSID\b\s*:\s*(.*?)\s*$")
# Match z.B. "Signal : 80%" - lokalisierungs-unabhaengig
_RE_SIGNAL_LINE = re.compile(r"(?m)^\s*\S[^:\n]*?:\s*(\d+)%\s*$")
# Scan-Block-Header "SSID 1 : foo"
_RE_SCAN_SSID = re.compile(r"(?m)^SSID\s+\d+\s*:\s*(.*?)\s*$")
# Profil-Liste: "Alle Benutzerprofile : MyNet" / "  All User Profile : MyNet"
# WICHTIG: alle Klassen mit \n verbieten, sonst spannt der Greedy-Match
# ueber mehrere Zeilen und faengt einen Doppelpunkt aus einer Folgezeile.
# Kein \b vor "profil" - im Deutschen wird das Wort in Komposita
# zusammengeschrieben ("Benutzerprofile") und haette dort keine Wortgrenze.
_RE_PROFILE_LINE = re.compile(
    r"(?im)^[^:\n]*profil[^:\n]*?:[ \t]*(.+?)[ \t]*$",
)


# --- Profil-XML-Generator ------------------------------------------------

_PROFILE_TEMPLATE_PSK = """<?xml version="1.0"?>
<WLANProfile xmlns="http://www.microsoft.com/networking/WLAN/profile/v1">
    <name>{ssid}</name>
    <SSIDConfig>
        <SSID>
            <name>{ssid}</name>
        </SSID>
    </SSIDConfig>
    <connectionType>ESS</connectionType>
    <connectionMode>auto</connectionMode>
    <MSM>
        <security>
            <authEncryption>
                <authentication>WPA2PSK</authentication>
                <encryption>AES</encryption>
                <useOneX>false</useOneX>
            </authEncryption>
            <sharedKey>
                <keyType>passPhrase</keyType>
                <protected>false</protected>
                <keyMaterial>{password}</keyMaterial>
            </sharedKey>
        </security>
    </MSM>
</WLANProfile>
"""

_PROFILE_TEMPLATE_OPEN = """<?xml version="1.0"?>
<WLANProfile xmlns="http://www.microsoft.com/networking/WLAN/profile/v1">
    <name>{ssid}</name>
    <SSIDConfig>
        <SSID>
            <name>{ssid}</name>
        </SSID>
    </SSIDConfig>
    <connectionType>ESS</connectionType>
    <connectionMode>auto</connectionMode>
    <MSM>
        <security>
            <authEncryption>
                <authentication>open</authentication>
                <encryption>none</encryption>
                <useOneX>false</useOneX>
            </authEncryption>
        </security>
    </MSM>
</WLANProfile>
"""


def _build_profile_xml(ssid: str, password: str) -> str:
    ssid_x = _xml_escape(ssid, quote=True)
    if not password:
        return _PROFILE_TEMPLATE_OPEN.format(ssid=ssid_x)
    pw_x = _xml_escape(password, quote=True)
    return _PROFILE_TEMPLATE_PSK.format(ssid=ssid_x, password=pw_x)


# --- Built-ins ----------------------------------------------------------

@builtin("WIFI_AVAILABLE", arity=0)
def _wifi_available():
    if not _IS_WINDOWS:
        return False
    try:
        rc, out = _run_netsh(["wlan", "show", "interfaces"])
    except GBRuntimeError:
        return False
    # Wenn kein WLAN-Adapter da ist, gibt netsh i.d.R. einen Hinweis aus
    # ohne SSID/Signal-Sektion; Returncode ist trotzdem 0. Wir checken
    # einfach ob das Schluesselwort 'WLAN' vorkommt.
    return rc == 0 and ("WLAN" in out or "Wireless" in out or "Drahtlos" in out)


@builtin("WIFI_CURRENT", arity=0)
def _wifi_current():
    _ensure_platform()
    _, out = _run_netsh(["wlan", "show", "interfaces"])
    # State/Status pruefen - nur 'connected'/'verbunden' liefert SSID
    state_line = ""
    for line in out.splitlines():
        ls = line.strip().lower()
        if (ls.startswith("state") or ls.startswith("status")
                or ls.startswith("zustand")) and ":" in ls:
            state_line = ls.split(":", 1)[1].strip()
            break
    connected = state_line in (
        "connected", "verbunden", "connecté", "connesso", "conectado"
    )
    if not connected:
        return ""
    m = _RE_SSID_LINE.search(out)
    return m.group(1).strip() if m else ""


@builtin("WIFI_SIGNAL", arity=0)
def _wifi_signal():
    _ensure_platform()
    _, out = _run_netsh(["wlan", "show", "interfaces"])
    m = _RE_SIGNAL_LINE.search(out)
    if not m:
        return -1
    try:
        return int(m.group(1))
    except ValueError:
        return -1


@builtin("WIFI_SCAN", arity=0)
def _wifi_scan():
    _ensure_platform()
    _, out = _run_netsh(["wlan", "show", "networks", "mode=bssid"])
    blocks = []
    cur_ssid = None
    cur_signal = None
    for line in out.splitlines():
        m = _RE_SCAN_SSID.match(line)
        if m:
            if cur_ssid is not None:
                blocks.append((cur_ssid, cur_signal))
            cur_ssid = m.group(1).strip() or "(versteckt)"
            cur_signal = None
            continue
        if cur_ssid is not None and cur_signal is None:
            sm = _RE_SIGNAL_LINE.match(line)
            if sm:
                try:
                    cur_signal = int(sm.group(1))
                except ValueError:
                    pass
    if cur_ssid is not None:
        blocks.append((cur_ssid, cur_signal))
    # nach Signal absteigend sortieren (None ans Ende)
    blocks.sort(key=lambda b: (b[1] is None, -(b[1] or 0)))
    return "\n".join(
        f"{s}|{sig if sig is not None else 0}" for s, sig in blocks
    )


@builtin("WIFI_CONNECT", arity=2, types=("str", "str"))
def _wifi_connect(ssid, password):
    _ensure_platform()
    if not ssid:
        raise GBRuntimeError("WIFI_CONNECT: SSID darf nicht leer sein")
    xml = _build_profile_xml(ssid, password)
    # Temp-Datei mit Profil-XML; netsh add profile filename=...
    tmp = Path(tempfile.gettempdir()) / f"_gb_wifi_{ssid}.xml"
    try:
        tmp.write_text(xml, encoding="utf-8")
        rc, out = _run_netsh([
            "wlan", "add", "profile",
            f"filename={tmp}", "user=current",
        ])
        if rc != 0:
            raise GBRuntimeError(
                f"WIFI_CONNECT: Profil konnte nicht angelegt werden\n{out.strip()}"
            )
    finally:
        try:
            tmp.unlink(missing_ok=True)
        except Exception:
            pass
    rc, out = _run_netsh(["wlan", "connect", f"name={ssid}"])
    return rc == 0


@builtin("WIFI_DISCONNECT", arity=0)
def _wifi_disconnect():
    _ensure_platform()
    rc, _ = _run_netsh(["wlan", "disconnect"])
    return rc == 0


@builtin("WIFI_PROFILES", arity=0)
def _wifi_profiles():
    _ensure_platform()
    _, out = _run_netsh(["wlan", "show", "profiles"])
    names = []
    seen = set()
    for m in _RE_PROFILE_LINE.finditer(out):
        n = m.group(1).strip()
        # Filter: Header-Zeilen "Gruppenrichtlinien..." enthalten zwar
        # "profil", haben aber keinen sinnvollen Wert nach dem Doppelpunkt.
        # Doppelte und group-prefix-Sektionen verwerfen.
        if n and n not in seen and not n.lower().startswith("group"):
            seen.add(n)
            names.append(n)
    return "\n".join(names)


@builtin("WIFI_DELETE_PROFILE", arity=1, types=("str",))
def _wifi_delete_profile(name):
    _ensure_platform()
    if not name:
        raise GBRuntimeError("WIFI_DELETE_PROFILE: Name darf nicht leer sein")
    rc, _ = _run_netsh(["wlan", "delete", "profile", f"name={name}"])
    return rc == 0

File: rust/gb_runtime/src/wifi.rs
//! WiFi-Management (WIFI_*) -- nativer Port von `gamebasic/modules/wifi.py`.
//! Windows-only via `netsh wlan` (std::process), Regex-Parsing. Feature `wifi`.
#![cfg(feature = "wifi")]

#[cfg(windows)]
mod imp {
    use std::os::windows::process::CommandExt;
    use std::process::Command;
    use std::time::Duration;

    const CREATE_NO_WINDOW: u32 = 0x0800_0000;
    /// Zeitlimit fuer `netsh`-Aufrufe -- ohne das kann ein haengender/
    /// ueberlasteter Netzwerk-Stack den kompletten Game-Loop-Thread
    /// unbegrenzt blockieren (`Command::output()` ist synchron/blockierend
    /// und laeuft auf demselben Thread wie Rendering/Input).
    const NETSH_TIMEOUT: Duration = Duration::from_secs(10);

    /// `netsh`-Konsolenausgabe ist auf Windows NICHT zuverlaessig eine feste
    /// Kodierung: je nach Windows-Version/Konfiguration liefert `netsh` bei
    /// umgeleitetem stdout (wie hier, `Command::output()`) mal die
    /// **OEM-Codepage** der Konsole (CP850/CP437/...), mal tatsaechlich UTF-8
    /// (empirisch auf diesem System bei `netsh wlan show profiles`
    /// verifiziert -- die vorherige Version, die blind ueber `GetOEMCP` +
    /// `MultiByteToWideChar` dekodierte, hat hier gueltiges UTF-8 kaputt
    /// gemacht statt es zu reparieren). Deshalb: erst UTF-8 versuchen (der
    /// haeufigere Fall bei umgeleitetem Output), nur bei ungueltigen Bytes auf
    /// die OEM-Codepage zurueckfallen (kein neues Crate, wie
    /// `local_datetime()` in builtins.rs -- rohe WinAPI-FFI).
    fn decode_console_output(bytes: &[u8]) -> String {
        if bytes.is_empty() {
            return String::new();
        }
        if let Ok(s) = std::str::from_utf8(bytes) {
            return s.to_string();
        }
        extern "system" {
            fn GetOEMCP() -> u32;
            fn MultiByteToWideChar(
                code_page: u32, dw_flags: u32,
                lp_multi_byte_str: *const u8, cb_multi_byte: i32,
                lp_wide_char_str: *mut u16, cch_wide_char: i32,
            ) -> i32;
        }
        unsafe {
            let cp = GetOEMCP();
            let needed = MultiByteToWideChar(cp, 0, bytes.as_ptr(), bytes.len() as i32, std::ptr::null_mut(), 0);
            if needed <= 0 {
                return String::from_utf8_lossy(bytes).into_owned();
            }
            let mut buf: Vec<u16> = vec![0u16; needed as usize];
            let written = MultiByteToWideChar(cp, 0, bytes.as_ptr(), bytes.len() as i32, buf.as_mut_ptr(), needed);
            if written <= 0 {
                return String::from_utf8_lossy(bytes).into_owned();
            }
            String::from_utf16_lossy(&buf[..written as usize])
        }
    }

    fn run_netsh(args: &[&str]) -> Result<(i32, String), String> {
        // `Command::output()` selbst hat kein Timeout -- auf einem Hilfs-Thread
        // laufen lassen und mit `recv_timeout` begrenzen (gleiches Muster wie
        // die DNS-Timeout-Aufloesung in net.rs).
        let args_owned: Vec<String> = args.iter().map(|s| s.to_string()).collect();
        let (tx, rx) = std::sync::mpsc::channel();
        std::thread::spawn(move || {
            let result = Command::new("netsh")
                .args(&args_owned)
                .creation_flags(CREATE_NO_WINDOW)
                .output();
            let _ = tx.send(result);
        });
        match rx.recv_timeout(NETSH_TIMEOUT) {
            Ok(Ok(out)) => {
                let mut text = decode_console_output(&out.stdout);
                text.push_str(&decode_console_output(&out.stderr));
                Ok((out.status.code().unwrap_or(-1), text))
            }
            Ok(Err(_)) => Err("WIFI: 'netsh' nicht gefunden".to_string()),
            Err(_) => Err(format!("WIFI: 'netsh' hat nach {}s nicht geantwortet", NETSH_TIMEOUT.as_secs())),
        }
    }

    fn re(pat: &str) -> regex::Regex {
        regex::Regex::new(pat).unwrap()
    }

    pub fn available() -> bool {
        // rc==0 heisst "netsh wlan show interfaces" konnte eine Schnittstelle
        // auflisten -- sprachunabhaengig und damit robuster als die vorherige
        // feste Woerterliste ("WLAN"/"Wireless"/"Drahtlos"), die auf jedem
        // Windows jenseits dieser drei Sprachen faelschlich FALSE lieferte
        // (enger als current()'s 5-Sprachen-Erkennung).
        matches!(run_netsh(&["wlan", "show", "interfaces"]), Ok((0, out)) if !out.trim().is_empty())
    }

    pub fn current() -> Result<String, String> {
        let (_, out) = run_netsh(&["wlan", "show", "interfaces"])?;
        let mut state = String::new();
        for line in out.lines() {
            let ls = line.trim().to_lowercase();
            if (ls.starts_with("state") || ls.starts_with("status") || ls.starts_with("zustand")) && ls.contains(':') {
                state = ls.split_once(':').unwrap().1.trim().to_string();
                break;
            }
        }
        let connected = matches!(state.as_str(), "connected" | "verbunden" | "connecté" | "connesso" | "conectado");
        if !connected {
            return Ok(String::new());
        }
        let rx = re(r"(?m)^\s*\bSSID\b\s*:\s*(.*?)\s*$");
        Ok(rx.captures(&out).and_then(|c| c.get(1)).map(|m| m.as_str().trim().to_string()).unwrap_or_default())
    }

    pub fn signal() -> Result<i64, String> {
        let (_, out) = run_netsh(&["wlan", "show", "interfaces"])?;
        let rx = re(r"(?m)^\s*\S[^:\n]*?:\s*(\d+)%\s*$");
        Ok(rx.captures(&out).and_then(|c| c.get(1)).and_then(|m| m.as_str().parse::<i64>().ok()).unwrap_or(-1))
    }

    pub fn scan() -> Result<String, String> {
        let (_, out) = run_netsh(&["wlan", "show", "networks", "mode=bssid"])?;
        let ssid_rx = re(r"(?m)^SSID\s+\d+\s*:\s*(.*?)\s*$");
        let sig_rx = re(r"(?m)^\s*\S[^:\n]*?:\s*(\d+)%\s*$");
        let mut blocks: Vec<(String, Option<i64>)> = Vec::new();
        let mut cur_ssid: Option<String> = None;
        let mut cur_sig: Option<i64> = None;
        for line in out.lines() {
            if let Some(c) = ssid_rx.captures(line) {
                if let Some(s) = cur_ssid.take() {
                    blocks.push((s, cur_sig.take()));
                }
                let name = c.get(1).map(|m| m.as_str().trim().to_string()).unwrap_or_default();
                cur_ssid = Some(if name.is_empty() { "(versteckt)".to_string() } else { name });
                cur_sig = None;
                continue;
            }
            if cur_ssid.is_some() && cur_sig.is_none() {
                if let Some(c) = sig_rx.captures(line) {
                    cur_sig = c.get(1).and_then(|m| m.as_str().parse::<i64>().ok());
                }
            }
        }
        if let Some(s) = cur_ssid.take() {
            blocks.push((s, cur_sig.take()));
        }
        blocks.sort_by_key(|b| (b.1.is_none(), -(b.1.unwrap_or(0))));
        Ok(blocks.iter().map(|(s, sig)| format!("{}|{}", s, sig.unwrap_or(0))).collect::<Vec<_>>().join("\n"))
    }

    fn xml_escape(s: &str) -> String {
        s.replace('&', "&amp;").replace('<', "&lt;").replace('>', "&gt;").replace('"', "&quot;").replace('\'', "&#x27;")
    }

    fn profile_xml(ssid: &str, password: &str) -> String {
        let ssid_x = xml_escape(ssid);
        if password.is_empty() {
            format!("<?xml version=\"1.0\"?>\n<WLANProfile xmlns=\"http://www.microsoft.com/networking/WLAN/profile/v1\">\n<name>{0}</name>\n<SSIDConfig><SSID><name>{0}</name></SSID></SSIDConfig>\n<connectionType>ESS</connectionType><connectionMode>auto</connectionMode>\n<MSM><security><authEncryption><authentication>open</authentication><encryption>none</encryption><useOneX>false</useOneX></authEncryption></security></MSM>\n</WLANProfile>\n", ssid_x)
        } else {
            format!("<?xml version=\"1.0\"?>\n<WLANProfile xmlns=\"http://www.microsoft.com/networking/WLAN/profile/v1\">\n<name>{0}</name>\n<SSIDConfig><SSID><name>{0}</name></SSID></SSIDConfig>\n<connectionType>ESS</connectionType><connectionMode>auto</connectionMode>\n<MSM><security><authEncryption><authentication>WPA2PSK</authentication><encryption>AES</encryption><useOneX>false</useOneX></authEncryption><sharedKey><keyType>passPhrase</keyType><protected>false</protected><keyMaterial>{1}</keyMaterial></sharedKey></security></MSM>\n</WLANProfile>\n", ssid_x, xml_escape(password))
        }
    }

    pub fn connect(ssid: &str, password: &str) -> Result<bool, String> {
        if ssid.is_empty() {
            return Err("WIFI_CONNECT: SSID darf nicht leer sein".into());
        }
        let xml = profile_xml(ssid, password);
        let mut tmp = std::env::temp_dir();
        tmp.push(format!("_gb_wifi_{}.xml", ssid.chars().filter(|c| c.is_ascii_alphanumeric()).collect::<String>()));
        std::fs::write(&tmp, xml).map_err(|e| format!("WIFI_CONNECT: temp-Datei: {}", e))?;
        let add = run_netsh(&["wlan", "add", "profile", &format!("filename={}", tmp.display()), "user=current"]);
        let _ = std::fs::remove_file(&tmp);
        let (rc, out) = add?;
        if rc != 0 {
            return Err(format!("WIFI_CONNECT: Profil konnte nicht angelegt werden\n{}", out.trim()));
        }
        let (rc2, _) = run_netsh(&["wlan", "connect", &format!("name={}", ssid)])?;
        Ok(rc2 == 0)
    }

    pub fn disconnect() -> Result<bool, String> {
        Ok(run_netsh(&["wlan", "disconnect"])?.0 == 0)
    }

    pub fn profiles() -> Result<String, String> {
        let (_, out) = run_netsh(&["wlan", "show", "profiles"])?;
        let rx = re(r"(?im)^[^:\n]*profil[^:\n]*?:[ \t]*(.+?)[ \t]*$");
        let mut seen = std::collections::HashSet::new();
        let mut names = Vec::new();
        for c in rx.captures_iter(&out) {
            let n = c.get(1).map(|m| m.as_str().trim().to_string()).unwrap_or_default();
            if !n.is_empty() && !n.to_lowercase().starts_with("group") && seen.insert(n.clone()) {
                names.push(n);
            }
        }
        Ok(names.join("\n"))
    }

    pub fn delete_profile(name: &str) -> Result<bool, String> {
        if name.is_empty() {
            return Err("WIFI_DELETE_PROFILE: Name darf nicht leer sein".into());
        }
        Ok(run_netsh(&["wlan", "delete", "profile", &format!("name={}", name)])?.0 == 0)
    }
}

#[cfg(not(windows))]
mod imp {
    fn unsupported<T>() -> Result<T, String> {
        Err("Modul WIFI: nur unter Windows unterstuetzt (nutzt netsh wlan).".into())
    }
    pub fn available() -> bool { false }
    pub fn current() -> Result<String, String> { unsupported() }
    pub fn signal() -> Result<i64, String> { unsupported() }
    pub fn scan() -> Result<String, String> { unsupported() }
    pub fn connect(_s: &str, _p: &str) -> Result<bool, String> { unsupported() }
    pub fn disconnect() -> Result<bool, String> { unsupported() }
    pub fn profiles() -> Result<String, String> { unsupported() }
    pub fn delete_profile(_n: &str) -> Result<bool, String> { unsupported() }
}

pub use imp::*;
